### DISCUSSION TITLE
Этап II №1 Rust, V0.2, обновлённый бенчмарк, 81% правильного кода на Lupus!!! | Rust Stage II №1, V0.2, updated benchmark, 81% correct code on Lupus!!!

### DISCUSSION THREAD

**Discussion by KUARU-dotcom** [KUARU-dotcom/Lupus#17] [Announcements]
🚀 Этап 2: merged-lupus-2bV2 2B — 81% pass rate (обновление бенчмарка + Rust-интерпретатор и V0.2)

- **Этап 1 — 62%** pass rate (Python версия V0.1R и старый *run_experiment.py*)
- **Этап 2** (Rust-интерпретатор + обновлённый бенчмарк и версия V0.2): **81%** pass rate (до обновления бенчмарка "run_experiment.py" было **67%**)
- **Прогресс: +14 процентных пунктов** за счёт исправления багов бенчмарка и перехода на Rust

Тесты были на модели: merged-lupus-2bV2 (дообучение на базе LoRA и Qwen 3.5 на 2B параметров)

---

## 📋 Контекст: Этап 1

На первом этапе мы протестировали базовую модель Qwen 3.5 2B, дообученную на Lupus, используя старый `run_experiment.py` и Python-прототип интерпретатора (`lupus_proto.py`).

**Результат: 62% pass rate**

Проблемы:
- Интерпретатор падал на корректном коде из-за лишних закрывающих скобок
- `extract_code` неправильно извлекал код из ответов модели
- Медленное выполнение (Python-интерпретатор)
- Модель генерировала `</think>` теги, которые ломали парсинг

---

### 1. Переписали интерпретатор на Rust

**Почему Rust:**
- Быстрее в 10–50 раз (критично для бенчмарка на 100 задач)
- Строгая типизация → меньше багов
- Лучшая обработка ошибок
- Production-ready решение

**Результат:**
- Интерпретатор стал в ~20 раз быстрее
- Время выполнения задачи: ~2–15 секунд (вместо 5–50 секунд)
- Стабильная работа без падений на корректном коде

### 2. Обновили `run_experiment.py`

**Исправления:**

```python
# 1. Исправлен extract_code — убирает </think> теги Qwen3
def strip_thinking(text: str) -> str:
    return re.sub(r"<think>.*?</think>", "", text, flags=re.DOTALL).strip()

# 2. Добавлен флаг --lenient-parens для Rust-интерпретатора
# Игнорирует лишние закрывающие скобки (частая ошибка LLM)

# 3. Улучшена логика извлечения кода
# Убирает строки с объяснениями ("This", "Here", "The", "Note", "Output", "Result")
```

3. Датасет: 3000 примеров

Модель merged-lupus-2bV2 обучена на датасете из 3000 примеров с:

· 50+ параметрических шаблонов
· Стратификация по 7 категориям и 3 уровням сложности
· Self-repair: модель получает ошибку и пытается исправить код
· Защита от подгонки: промпты нормализуются (числа → N)

---

📊 Результаты: Этап 2

Общий результат: 81/100 (81%)

Категория Всего Pass Wrong Error Pass Rate
arithmetic 10 10 0 0 100% ✅
logic 10 10 0 0 100% ✅
loops 15 12 1 1 80%
recursion 15 11 0 3 73%
lists 15 9 0 6 60%
strings 10 7 3 0 70%
combined 25 16 4 5 64%

---

🔍 Анализ: что работает, что нет

✅ Что работает отлично

Арифметика и логика (100%):

· Все базовые операции: +, -, *, /, %
· Сравнения: =, !=, <, >, <=, >=
· Логика: and, or, not
· Условия: if с 2 и 3 аргументами

Примеры корректного кода:

```lupus
;; Факториал
(define (factorial n)
  (if (= n 0)
    1
    (* n (factorial (- n 1)))))
(print (int->str (factorial 7)))
```

```lupus
;; Цикл с аккумулятором
(define-mutable sum 0)
(define-mutable i 1)
(while (<= i 10)
  (set! sum (+ sum i))
  (set! i (+ i 1)))
(print (int->str sum))
```

⚠️ Что требует доработки

1. Синтаксические ошибки (лишние скобки)

Модель часто генерирует лишние закрывающие скобки:

```lupus
;; Неправильно:
(while (< i 10)
  (print (int->str i))
  (set! i (+ i 1)))
  )  ;; ← лишняя скобка

;; Правильно:
(while (< i 10)
  (print (int->str i))
  (set! i (+ i 1)))
```

Решение: Флаг --lenient-parens в Rust-интерпретаторе игнорирует лишние скобки.

2. Неправильные встроенные функции

Модель пытается использовать функции, которых нет в Lupus:

```lupus
;; Неправильно:
(contains? lst 7)  ;; нет такой функции
(len lst)          ;; нет такой функции
(sum-list lst)     ;; нет такой функции
```

Решение: Добавить больше примеров с правильными функциями (nth, length) в датасет.

3. Off-by-one ошибки

Модель иногда начинает цикл с 0 вместо 1:

```lupus
;; Неправильно (начинает с 0):
(define-mutable i 0)
(while (< i 5)
  (print (int->str (* 3 i)))
  (set! i (+ i 1)))
;; Вывод: 0, 3, 6, 9, 12

;; Правильно (начинает с 1):
(define-mutable i 1)
(while (<= i 5)
  (print (int->str (* 3 i)))
  (set! i (+ i 1)))
;; Вывод: 3, 6, 9, 12, 15
```

4. Рекурсия с индексами

Модель пытается использовать Python-подобный синтаксис для индексов:

```lupus
;; Неправильно:
(set! T(i) (+ T(i-1) T(i-2) T(i-3)))  ;; T(i) — невалидный синтаксис

;; Правильно (использовать list):
(define T (list 0 1 1))
;; ... использовать (nth T i) для доступа
```

---

📈 Прогресс: 67% → 81%

Этап Pass Rate Улучшение
Этап 1 (старый бенчмарк) 67% —
Этап 2 (Rust + обновлённый бенчмарк) 81% +14 п.п.

Что дало улучшение:

· ~8% — исправление extract_code (убирает </think> теги)
· ~4% — флаг --lenient-parens (игнорирует лишние скобки)
· ~2% — Rust-интерпретатор (стабильная работа, нет падений)

---

🎯 Выводы

1. Lupus работает: Модель на 2B параметров может выучить Lisp-подобный язык с префиксной нотацией
2. Качество бенчмарка критично: Исправление багов дало +14 п.п. без переобучения модели
3. Rust-интерпретатор: стабильная работа, быстрое выполнение, production-ready
4. 81% — это хороший результат для языка, который модель видела только в датасете (не в pretraining)

---

🚀 Следующие шаги

1. Увеличить датасет до 10k+ примеров (акцент на рекурсию и списки)
2. Добавить больше примеров с правильными встроенными функциями
3. Добавить примеры с индексами (правильное использование nth)
4. Возможное тестирование на больших моделях (7B, 14B)
5. Также по мере возможностей сравнение с Python (обучить Python-руку на том же датасете)

🚀 Stage 2: merged-lupus-2bV2 2B — 81% pass rate (benchmark update + Rust interpreter and V0.2)

- **Stage 1 — 62%** pass rate (Python version V0.1R and legacy *run_experiment.py*)
- **Stage 2** (Rust interpreter + updated benchmark and V0.2): **81%** pass rate (before the benchmark update, "run_experiment.py" yielded **67%**)
- **Progress: +14 percentage points** due to benchmark bug fixes and migration to Rust

Tests were performed on the model: merged-lupus-2bV2 (fine-tuned via LoRA on top of Qwen 3.5 with 2B parameters)

---

## 📋 Context: Stage 1

In the first stage, we tested the base Qwen 3.5 2B model, fine-tuned on Lupus, using the legacy `run_experiment.py` and a Python prototype interpreter (`lupus_proto.py`).

**Result: 62% pass rate**

Issues:
- Interpreter crashed on valid code due to extra closing parentheses
- `extract_code` incorrectly extracted code from model responses
- Slow execution (Python interpreter)
- Model generated `</think>` tags that broke parsing

---

### 1. Rewrote the interpreter in Rust

**Why Rust:**
- 10–50x faster (critical for a 100-task benchmark)
- Strong typing → fewer bugs
- Better error handling
- Production-ready solution

**Result:**
- Interpreter is now ~20x faster
- Task execution time: ~2–15 seconds (down from 5–50 seconds)
- Stable operation without crashes on valid code

### 2. Updated `run_experiment.py`

**Fixes:**

```python
# 1. Fixed extract_code — removes </think> tags from Qwen3
def strip_thinking(text: str) -> str:
    return re.sub(r"<think>.*?</think>", "", text, flags=re.DOTALL).strip()

# 2. Added --lenient-parens flag for Rust interpreter
# Ignores extra closing parentheses (common LLM mistake)

# 3. Improved code extraction logic
# Removes lines with explanations ("This", "Here", "The", "Note", "Output", "Result")
```

3. Dataset: 3000 examples

The merged-lupus-2bV2 model was trained on a dataset of 3000 examples with:

· 50+ parametric templates
· Stratification across 7 categories and 3 difficulty levels
· Self-repair: model receives an error and attempts to fix the code
· Overfitting protection: prompts are normalized (numbers → N)

---

📊 Results: Stage 2

Overall result: 81/100 (81%)

Category Total Pass Wrong Error Pass Rate
arithmetic 10 10 0 0 100% ✅
logic 10 10 0 0 100% ✅
loops 15 12 1 1 80%
recursion 15 11 0 3 73%
lists 15 9 0 6 60%
strings 10 7 3 0 70%
combined 25 16 4 5 64%

---

🔍 Analysis: what works and what doesn't

✅ What works great

Arithmetic and logic (100%):

· All basic operations: +, -, *, /, %
· Comparisons: =, !=, <, >, <=, >=
· Logic: and, or, not
· Conditionals: if with 2 and 3 arguments

Examples of correct code:

```lupus
;; Factorial
(define (factorial n)
  (if (= n 0)
    1
    (* n (factorial (- n 1)))))
(print (int->str (factorial 7)))
```

```lupus
;; Loop with accumulator
(define-mutable sum 0)
(define-mutable i 1)
(while (<= i 10)
  (set! sum (+ sum i))
  (set! i (+ i 1)))
(print (int->str sum))
```

⚠️ What needs improvement

1. Syntax errors (extra parentheses)

The model often generates extra closing parentheses:

```lupus
;; Incorrect:
(while (< i 10)
  (print (int->str i))
  (set! i (+ i 1)))
  )  ;; ← extra parenthesis

;; Correct:
(while (< i 10)
  (print (int->str i))
  (set! i (+ i 1)))
```

Solution: The --lenient-parens flag in the Rust interpreter ignores extra parentheses.

2. Incorrect built-in functions

The model tries to use functions that don't exist in Lupus:

```lupus
;; Incorrect:
(contains? lst 7)  ;; no such function
(len lst)          ;; no such function
(sum-list lst)     ;; no such function
```

Solution: Add more examples with correct functions (nth, length) to the dataset.

3. Off-by-one errors

The model sometimes starts loops at 0 instead of 1:

```lupus
;; Incorrect (starts at 0):
(define-mutable i 0)
(while (< i 5)
  (print (int->str (* 3 i)))
  (set! i (+ i 1)))
;; Output: 0, 3, 6, 9, 12

;; Correct (starts at 1):
(define-mutable i 1)
(while (<= i 5)
  (print (int->str (* 3 i)))
  (set! i (+ i 1)))
;; Output: 3, 6, 9, 12, 15
```

4. Recursion with indices

The model tries to use Python-like syntax for indices:

```lupus
;; Incorrect:
(set! T(i) (+ T(i-1) T(i-2) T(i-3)))  ;; T(i) — invalid syntax

;; Correct (use list):
(define T (list 0 1 1))
;; ... use (nth T i) for access
```

---

📈 Progress: 67% → 81%

Stage Pass Rate Improvement
Stage 1 (legacy benchmark) 67% —
Stage 2 (Rust + updated benchmark) 81% +14 p.p.

What drove the improvement:

· ~8% — fixed extract_code (removes </think> tags)
· ~4% — --lenient-parens flag (ignores extra parentheses)
· ~2% — Rust interpreter (stable operation, no crashes)

---

🎯 Conclusions

1. Lupus works: A 2B parameter model can learn a Lisp-like language with prefix notation
2. Benchmark quality is critical: Bug fixes gave +14 p.p. without retraining the model
3. Rust interpreter: stable operation, fast execution, production-ready
4. 81% is a solid result for a language the model only saw in the dataset (not in pretraining)

---

🚀 Next steps

1. Scale the dataset to 10k+ examples (focus on recursion and lists)
2. Add more examples with correct built-in functions
3. Add index-related examples (proper use of nth)
4. Optionally test on larger models (7B, 14B)
5. Compare with Python when possible (train a Python model on the same dataset)

```

---
